### PR TITLE
fix the ParallelExecutor on Windows

### PR DESCRIPTION
**Pull request by wopeizl**
test=develop

### CODE REVIEW

File: paddle/fluid/framework/parallel_executor.cc [hunk 1]
```diff
@@ -290,6 +290,14 @@ ParallelExecutor::ParallelExecutor(const std::vector<platform::Place> &places,
...
       build_strategy.reduce_ == BuildStrategy::ReduceStrategy::kAllReduce;
   member_->nranks_ = build_strategy.num_trainers_ * places.size();
+#if defined(PADDLE_WITH_CUDA) && defined(_WIN32)
+  if (!member_->use_all_reduce_) {
+    LOG_FIRST_N(WARNING, 1)
+        << "Please NOTE: Windows can support single GPU only"
+        << ", parallelExecutor would be forced to use allreduce.";
+    member_->use_all_reduce_ = true;
+  }
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:300]: Why do you add those?
Currently, it is not supported that using multi-GPU to train or test, so you should check the number of places first.

File: paddle/fluid/framework/parallel_executor.cc [hunk 2]
```diff
@@ -336,7 +344,8 @@ ParallelExecutor::ParallelExecutor(const std::vector<platform::Place> &places,
 
   if (member_->use_cuda_) {
 // Bcast Parameters to all GPUs
-#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
+#if defined(PADDLE_WITH_CUDA)
+#if !defined(_WIN32)
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:348]: There should not have so many `if`, it is ugly.

File: paddle/fluid/framework/parallel_executor.cc [hunk 3]
```diff
@@ -506,7 +514,8 @@ void ParallelExecutor::BCastParamsToDevices(
     }
     auto &dims = main_tensor.dims();
     if (paddle::platform::is_gpu_place(main_tensor.place())) {
-#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
+#if defined(PADDLE_WITH_CUDA)
+#if !defined(_WIN32)
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:518]: Same above.

File: paddle/fluid/framework/parallel_executor.cc [hunk 4]
```diff
@@ -653,11 +663,20 @@ bool ParallelExecutor::EnableParallelGraphExecution(
...
+  if (!member_->use_all_reduce_ || !member_->use_cuda_) {
     if (build_strategy.enable_sequential_execution_ ||
-        exec_strategy.type_ == ExecutionStrategy::ExecutorType::kExperimental)
+        exec_strategy.type_ == ExecutionStrategy::ExecutorType::kExperimental) {
       enable_parallel_graph = false;
+    }
+  }
+
+#ifdef WIN32
+  LOG_FIRST_N(WARNING, 1)
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:674]: Why do you use `LOG_FIRST_N ` here?

File: paddle/fluid/pybind/pybind.cc [hunk 5]
```diff
@@ -1245,6 +1245,12 @@ All parameter, weight, gradient are variables in Paddle.
           [](BuildStrategy &self, BuildStrategy::ReduceStrategy strategy) {
             PADDLE_ENFORCE(!self.IsFinalized(), "BuildStrategy is finlaized.");
             self.reduce_ = strategy;
+#ifdef WIN32
+            if (self.reduce_ == BuildStrategy::ReduceStrategy::kReduce) {
+              LOG_FIRST_N(WARNING, 1) << "Windows support AllReduce only.";
```
Comment (chengduoZH) [inline paddle/fluid/pybind/pybind.cc:1250]: Do you know the difference between of Reduce and AllReduce mode?

Comment (chengduoZH) [inline paddle/fluid/pybind/pybind.cc:1250]: Why use `LOG_FIRST_N ` here?

File: python/paddle/fluid/framework.py [hunk 6]
```diff
@@ -37,6 +37,13 @@
             __file__)) + os.sep + '..' + os.sep + 'libs'
         os.environ['path'] += ';' + third_lib_path
         sys.path.append(third_lib_path)
+        # Windows can support single GPU only
+        if hasattr(os.environ, 'CUDA_VISIBLE_DEVICES') and \
+                os.environ['CUDA_VISIBLE_DEVICES'] is not None:
+            os.environ['CUDA_VISIBLE_DEVICES'] = os.environ[
+                'CUDA_VISIBLE_DEVICES'].strip().split(',')[0]
+        else:
+            os.environ['CUDA_VISIBLE_DEVICES'] = '0'
```
Comment (chengduoZH) [inline python/paddle/fluid/framework.py:46]: Why you specify the `CUDA_VISIBLE_DEVICES`? We should not modify user configuration information without authorization.

File: python/paddle/fluid/tests/unittests/test_parallel_executor.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_parallel_executor.py:1]: ->2019.

Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_parallel_executor.py:1]: The test name, `test_parallel_executor.py`, is confusing.
Why do you add `test_parallel_executor.py`?

File: python/paddle/fluid/tests/unittests/test_parallel_executor.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    hidden = fluid.layers.fc(
+        img,
+        size=200,
+        act='tanh',
+        bias_attr=fluid.ParamAttr(
+            initializer=fluid.initializer.Constant(value=1.0)))
+    prediction = fluid.layers.fc(hidden, size=10, act='softmax')
+    loss = fluid.layers.cross_entropy(input=prediction, label=label)
+    loss = fluid.layers.mean(loss)
+    return loss
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_parallel_executor.py:35]: `simple_net.py` has defined `simple_fc_net`, you can use it directly.

File: python/paddle/fluid/tests/unittests/test_parallel_executor.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+
+class TestParallelExecutorStrategy(TestParallelExecutorBase):
+    def _init_data(self):
+        np.random.seed(5)
+        img = np.random.random(size=[32, 784]).astype(np.float32)
+        label = np.ones(shape=[32, 1], dtype='int64')
+        return img, label
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_parallel_executor.py:43]: You can use `init_data` of `simple_net.py`.

Comment (wopeizl) [reply python/paddle/fluid/tests/unittests/test_parallel_executor.py:43]: ok

File: python/paddle/fluid/tests/unittests/test_parallel_executor.py (hunk 8 above)
Comment (wopeizl) [reply python/paddle/fluid/tests/unittests/test_parallel_executor.py:35]: ok

File: python/paddle/fluid/tests/unittests/test_parallel_executor.py (hunk 7 above)
Comment (wopeizl) [reply python/paddle/fluid/tests/unittests/test_parallel_executor.py:1]: Just for the coverage test purpose, I didnt find a file cover all the flags regarding to parallel_executor.

Comment (wopeizl) [reply python/paddle/fluid/tests/unittests/test_parallel_executor.py:1]: thanks

File: python/paddle/fluid/framework.py (hunk 6 above)
Comment (wopeizl) [reply python/paddle/fluid/framework.py:46]: Thanks, removed

File: paddle/fluid/pybind/pybind.cc (hunk 5 above)
Comment (wopeizl) [reply paddle/fluid/pybind/pybind.cc:1250]: Got it

File: paddle/fluid/framework/parallel_executor.cc (hunk 3 above)
Comment (wopeizl) [reply paddle/fluid/framework/parallel_executor.cc:518]: Cleaned.

File: paddle/fluid/framework/parallel_executor.cc (hunk 2 above)
Comment (wopeizl) [reply paddle/fluid/framework/parallel_executor.cc:348]: Cleaned.

File: paddle/fluid/framework/parallel_executor.cc (hunk 1 above)
Comment (wopeizl) [reply paddle/fluid/framework/parallel_executor.cc:300]: ok

File: paddle/fluid/framework/parallel_executor.cc (hunk 4 above)
Comment (wopeizl) [reply paddle/fluid/framework/parallel_executor.cc:674]: Dont hope this bother users all the time.

**Comment by wopeizl** [comment]
https://github.com/PaddlePaddle/Paddle/pull/17787

**Pull request closed without merging**